doc/design.md

```diff
@@ -47,6 +61,53 @@ subset of all the functinoality and it requires all our objects to have a
+
+### Conclusion:
+
+* We should pass around `Ptr` instead of `RCP` whenever possible, and we can do this right away. 
+* Use `UniquePtr` (after it is implemented) whenever possible instead of `RCP` --- though most places in `SymEngine` seem to require `RCP`. But we should keep this in mind, there might still be one or two places where `UniquePtr` is the way to go.
+* Never use raw new/delete and never use raw pointers (use `UniquePtr` + `Ptr`, and if it is not sufficient, use the slower `RCP` + `Ptr`).
+* Never pass pointers to some internal data --- pass `Ptr` and have it Debug time checked by using `UniquePtr` in Debug mode, and use the data directly in Release mode
+
+As an example of the last point, e.g. to give access to an internal `std::map` (as a pointer, so that the map is not copied), here is how to do it:
+```c++
+class A {
+private:
+#ifdef DEBUG_MODE
+    UniquePtr<std::map<int, int> > m;
+#else
+    std::map<int, int> m;
+#endif
+public:
+    Ptr<std::map<int, int>> get_access() {
+#ifdef DEBUG_MODE
+        return m.ptr();
+#else
+        return ptrFromRef(m);
+#endif
+    }
+};
```

Comment:
Note that this is easier done using the `Viewable` class in https://github.com/certik/trilinos/pull/2, but we can update the docs once I implement it.